### PR TITLE
ITT-141 Automating Services in Java Lib

### PR DESCRIPTION
**Pull request by jillingk**
**Description**
Created this draft PR to show progress on the java service automation.
TODO: 
- baseURL formatting (test/live) -> done
- Overloading methods for non required params -> done
- Create single file service -> done
- Method description -> done

### CODE REVIEW

**Comment by wboereboom** [comment]
Should we remove the existing service? Or is it better to deprecate first?

**Comment by jillingk** [comment]
> Should we remove the existing service? Or is it better to deprecate first?

@wboereboom difficult to decide tbh. Would be easier to migrate if we deprecate it, but then again we're cluttering the lib quite a lot..

File: Makefile [hunk 1]
```diff
@@ -52,6 +59,62 @@ $(services): target/spec $(openapi-generator-jar)
...
+		-g $(generator) \
+		-t templates \
+		-o $(output) \
+		--reserved-words-mappings configuration=configuration \
+		--ignore-file-override ./.openapi-generator-ignore \
+		--skip-validate-spec \
+		--model-package $(subst /,.,com.adyen.model.$@) \
+		--library $(library) \
+		--api-package com.adyen.service.$@ \
+        --api-name-suffix Service \
```
Comment (michaelpaul) [inline Makefile:82]: I kind like the `Api` suffix, it produces classes like `PaymentsApi` and `ModificationsApi`. Checkout as a whole is a service and these classes are parts of the whole, but not a service on their own. They also form the main API offered by the library to use our REST API's, seems like a good choice of word. ;) 

```suggestion
        --api-name-suffix Api \
```

File: Makefile [hunk 2]
```diff
@@ -52,6 +59,62 @@ $(services): target/spec $(openapi-generator-jar)
 	mv $(output)/$(models)/$@ $(models)/$@
 	mv $(output)/$(models)/JSON.java $(models)/$@
 
+# Full service + models automation
+bigServices:=balanceplatform checkout storedValue payments payout management legalentitymanagement transfers
+singleFileServices:=balancecontrol binlookup dataprotection storedvalue posterminalmanagement recurring
+
+all: $(bigServices) $(singleFileServices)
```
Comment (michaelpaul) [inline Makefile:66]: The `all` target is invoked when you run `make` (without any target). I would rather to have two targets, `models` for when generating only models and `services` for both models and service classes.

```suggestion
services: $(bigServices) $(singleFileServices)
```

**Review by michaelpaul — APPROVED** [review]
Great job @jillingk! 🚀

**Comment by michaelpaul** [comment]
The way query string parameters are passed is changing from `Map<String, String>` to actual method arguments. The order of path, body, and query has also changed. 

This change is good and can be done to existing services (e.g. Management and Balance Platform) in a new major release.

File: Makefile [hunk 3]
```diff
@@ -52,6 +59,62 @@ $(services): target/spec $(openapi-generator-jar)
...
+		-g $(generator) \
+		-t templates \
+		-o $(output) \
+		--reserved-words-mappings configuration=configuration \
+		--ignore-file-override ./.openapi-generator-ignore \
+		--skip-validate-spec \
+		--model-package $(subst /,.,com.adyen.model.$@) \
+		--library $(library) \
+		--api-package com.adyen.service.$@ \
+        --api-name-suffix Api \
```
Comment (wboereboom) [inline Makefile:82]:
```suggestion
		--api-name-suffix Api \
```

File: Makefile [hunk 4]
```diff
@@ -52,6 +59,62 @@ $(services): target/spec $(openapi-generator-jar)
...
+		-c templates/libraries/okhttp-gson/config.yaml \
+		-o $(output) \
+		--reserved-words-mappings configuration=configuration \
+		--ignore-file-override ./.openapi-generator-ignore \
+		--skip-validate-spec \
+		--model-package $(subst /,.,com.adyen.model.$@) \
+		--library $(library) \
+		--additional-properties customApi=$@ \
+		--api-package com.adyen.service \
+        --api-name-suffix Api \
```
Comment (wboereboom) [inline Makefile:107]:
```suggestion
		--api-name-suffix Api \
```

File: src/main/java/com/adyen/Service.java [hunk 5]
```diff
@@ -49,4 +50,27 @@ public void setApiKeyRequired(boolean apiKeyRequired) {
         isApiKeyRequired = apiKeyRequired;
     }
 
+    public String createBaseURL(String url) {
```
Comment (jillingk) [inline src/main/java/com/adyen/Service.java:53]:
```suggestion
    protected String createBaseURL(String url) {
```

**Review by michaelpaul — APPROVED** [review]

**Review by antolo-arch — APPROVED** [review]

**Review by AlexandrosMor — APPROVED** [review]

**Pull request merged**